File: src/main/java/org/nia/logic/quests/ICrossQuestStep.java
package org.nia.logic.quests;

import org.nia.model.Quest;
import org.nia.model.QuestEvent;

import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

/**
 * @author IANazarov
 */
public interface ICrossQuestStep extends IQuestStep {

    @Override
    default String getGoodText(Quest quest) {
        return getGoodInactiveText();
    }

    default String getGoodInactiveText() {
        return "Твой коллега ничего не предпринял, так что ты выставил его крайним и смотался.";
    }

    default String getBadInactiveText() {
        return "Ты ничего не предпринял и всё пошло наперекосяк.";
    }

    @Override
    default String getBadText(Quest quest) {
        return getBadInactiveText();
    }

    @Override
    default String getText(Quest quest) {
        return "";
    }

    default boolean isWaitUser() {
        return false;
    }

    default boolean isButtonWithUser() {
        return false;
    }

    List<IQuestAction> getActionList();

    default IQuestAction getProducedAction(QuestEvent event) {
        List<IQuestAction> actionList = getActionList();
        if (actionList.isEmpty()) {
            return null;
        } else {
            boolean isFinal = !isWaitUser() && getNext(event.getQuest()).isEmpty();
            if (!isFinal) {
                return actionList.get(new Random().nextInt(actionList.size()));
            } else if (isWin(event)) {
                List<IQuestAction> winList = actionList.stream()
                        .filter(IQuestAction::isWin)
                        .collect(Collectors.toList());
                if (winList.isEmpty()) {
                    return null;
                } else {
                    return winList.get(new Random().nextInt(winList.size()));
                }
            } else {
                List<IQuestAction> loseList = actionList.stream()
                        .filter(IQuestAction::isLose)
                        .collect(Collectors.toList());
                if (loseList.isEmpty()) {
                    return null;
                } else {
                    return loseList.get(new Random().nextInt(loseList.size()));
                }
            }
        }
    }
}

File: src/main/java/ru/nia/tavern/quests/ICrossQuestStep.kt
package ru.nia.tavern.quests

import ru.nia.tavern.model.Quest
import ru.nia.tavern.model.QuestEvent
import java.util.*
import java.util.stream.Collectors

/**
 * @author IANazarov
 */
interface ICrossQuestStep : IQuestStep {

    val goodInactiveText: String
        get() = "Твой коллега ничего не предпринял, так что ты выставил его крайним и смотался."

    val badInactiveText: String
        get() = "Ты ничего не предпринял и всё пошло наперекосяк."

    val isWaitUser: Boolean
        get() = false

    val isButtonWithUser: Boolean
        get() = false

    val actionList: List<IQuestAction>

    override fun getGoodText(quest: Quest): String {
        return goodInactiveText
    }

    override fun getBadText(quest: Quest): String {
        return badInactiveText
    }

    override fun getText(quest: Quest): String {
        return ""
    }

    fun getProducedAction(event: QuestEvent): IQuestAction? {
        val actionList = actionList
        if (actionList.isEmpty()) {
            return null
        } else {
            val isFinal = !isWaitUser && getNext(event.quest).isEmpty()
            if (!isFinal) {
                return actionList[Random().nextInt(actionList.size)]
            } else if (isWin(event)) {
                val winList = actionList.stream()
                        .filter(Predicate<IQuestAction> { it.isWin() })
                        .collect<List<IQuestAction>, Any>(Collectors.toList())
                return if (winList.isEmpty()) {
                    null
                } else {
                    winList[Random().nextInt(winList.size)]
                }
            } else {
                val loseList = actionList.stream()
                        .filter(Predicate<IQuestAction> { it.isLose() })
                        .collect<List<IQuestAction>, Any>(Collectors.toList())
                return if (loseList.isEmpty()) {
                    null
                } else {
                    loseList[Random().nextInt(loseList.size)]
                }
            }
        }
    }
}
